### DISCUSSION TITLE
[Question] A way to create a virtualized grid?

### DISCUSSION THREAD

**Discussion by jvulcan** [AvaloniaUI/Avalonia#6659] [Q&A] [answered]
Hi, I'm trying to virtualize my grid because when you add too much elements the FPS drops heavily (around 4 FPS). It is some kind of table
![grid_example](https://user-images.githubusercontent.com/581282/135527645-0c08c49b-d386-4c4e-94d8-932c42593c26.png)

AS you can see for every combination of ALT & CRIT, a control must be created, so when you add 50+ Alts and 50+ Crits, the amount of controls is so high that the FPS tanks severely and naturally, memory jumps up, but I suppose that is expected.

I saw the ListBox can Virtualize Items, but I need a Grid. I could make a ListBox for every CRIT, but I would have 50+ ListBoxes so the FPS would tank anyway if there are too many CRITs,
The other issue is that I would want a ScrollBar in there, if I move it vertically, every 'listbox' would displace along but also the ALT list, and the same with CRIT and moving horizontally.

In summary this is like an Excel grid,

A four blocks grid is my starting base
`

<Grid 
         ColumnDefinitions="200,*" 
         RowDefinitions="150,*" 
         ShowGridLines="True">

              
         <TreeView 
                 Name="tvAlts"
                 Grid.Row="1" Grid.Column="0"
                 Background="BlanchedAlmond">
                 <TreeView.ItemTemplate>
                         <DataTemplate DataType="local:Alternative">
                                 <Label 
                                        Content="{Binding Name}" 
                                        VerticalContentAlignment="Center"
                                        Height="40"
                                        Padding="0"
                                        Margin="0,-1,-1,-1"/>
                         </DataTemplate>
                 </TreeView.ItemTemplate>
         </TreeView>

         <ScrollViewer 
                 Grid.Column="1" 
                 VerticalScrollBarVisibility="Disabled" 
                 HorizontalScrollBarVisibility="Auto">
                 <WrapPanel
                         Name="wpCrits"
                         Orientation="Horizontal"
                         Grid.Column="1" Grid.Row="0"
                         Background="WhiteSmoke">
                 </WrapPanel>
         </ScrollViewer>

         <ScrollViewer Grid.Column="1" Grid.Row="1">
                 <ItemsPresenter 
                           Grid.Column="1"
                           Grid.Row="1"                                                   
                           Width="200"
                           HorizontalAlignment="Left"
                           VirtualizationMode="Simple" 
                           Name="ipTest">

                           <ItemsPresenter.ItemTemplate>
                                     <DataTemplate DataType="local:Alternative">
                                              <ItemsPresenter
                                                      ScrollViewer.HorizontalScrollBarVisibility="Visible"
                                                      Items="{Binding LLCriteria}"
                                                      HorizontalAlignment="Left"
                                                      VirtualizationMode="Simple">
                                                      <ItemsPresenter.ItemsPanel>
                                                                <ItemsPanelTemplate>
                                                                        <StackPanel Orientation="Horizontal"/>
                                                                </ItemsPanelTemplate>
                                                      </ItemsPresenter.ItemsPanel>
                                                      <ItemsPresenter.ItemTemplate>
                                                                      <DataTemplate>
                                                                                <Label
                                                                                         Width="200"
                                                                                         Content="{Binding}"
                                                                                         Height="40"
                                                                                         HorizontalContentAlignment="Center"
                                                                                         VerticalContentAlignment="Center"/>
                                                                      </DataTemplate>
                                                      </ItemsPresenter.ItemTemplate>
                                              </ItemsPresenter>
                                     </DataTemplate>
                           </ItemsPresenter.ItemTemplate>
                 </ItemsPresenter>
         </ScrollViewer>
</Grid>

`
The custom Horizontal stackpanel looks like it's not helping with virtualization.

I have other challenges but I prefer to just focus in this one that impacts the experience more heavily.

Thanks in advance.

**Comment by maxkatz6** [comment, marked as answer]
Vertical virtualized ItemsRepeater for rows with other ItemsRepeater inside of each row for cells would worth a try.
ItemsRepeater with UniformGridLayout can also just work, in fact I would try this approach at first. There is an example in controls catalog, but API is pretty much the same as in UWP.

But for even more optimized solution better to ask @grokys

**Reply by jvulcan** [reply]
Thanks for the reply 👌, 
I'm going to take a look at the UniformGridLayout tomorrow and then reply again about my testings

**Reply by jvulcan** [reply]
Hi again, here are my testing results with ItemsRepeater + UniformGridLayout:

1. It looks like virtualization doesn't work with ItemsRepeater? I mean, there's no VirtualizationMode property to begin with.
2. It works with 20x20 elements, but beyond that many elements are not drawn (testing with 40x40 elements):

https://user-images.githubusercontent.com/581282/135645429-ad2dfd6b-8445-4f6b-8a6d-1c0652505ddd.mp4

3. It even crashes while scrolling horizontally:

https://user-images.githubusercontent.com/581282/135645530-b09c5366-093b-4dbb-ae90-734538d78d9d.mp4

Here's an extract of my XAML code of the inner portion:

`
    <ScrollViewer 
        Name="innerScroll"
        Grid.Column="1" Grid.Row="1" 
        HorizontalScrollBarVisibility="Auto" 
        VerticalScrollBarVisibility="Auto">
        
        <ItemsRepeater 
            HorizontalAlignment="Left"
            Name="irTest"
            Items="{Binding EveryEval}">
            <ItemsRepeater.Layout>
	        <UniformGridLayout 
                    MinItemWidth="200" 
                    MinItemHeight="40"
                    Orientation="Vertical"/>
            </ItemsRepeater.Layout>
            
            <ItemsRepeater.ItemTemplate>
	        <DataTemplate>
                    <TextBox 
                        Text="{Binding}"
                        Margin="2"
                        HorizontalContentAlignment="Center"
                        VerticalContentAlignment="Center"/>
	        </DataTemplate>
            </ItemsRepeater.ItemTemplate>
        </ItemsRepeater>
        
    </ScrollViewer>
`

This will get worse because a Textbox is just for testing, I need to use several controls in each grid element.
And like a pointed out earlier, it looks like ItemsRepeater doesn't virtualize, because while virtualizing the scrollviewer behaves different, it jumps (IIRC). Maybe it's my own scrollviewer that's messing up?

Awaiting for your reply.
Thanks in advance.

**Reply by maxkatz6** [reply]
UniformGridLayout might be unfortunately buggy for these complex cases with thousands of cells.

> It looks like virtualization doesn't work with ItemsRepeater? I mean, there's no VirtualizationMode property to begin with.

It always virtualizes, when it can. 

> Maybe it's my own scrollviewer that's messing up?

Your "own"? Is it custom? Or do you just mean that you put it there instead of it being part of the template (like in the ListBox)? It is supposed to be used like this - ItemsRepeater inside of ScrollViewer.

**Reply by jvulcan** [reply]
Thank you for the reply.

> Your "own"? Is it custom? Or do you just mean that you put it there instead of it being part of the template (like in the ListBox)? It is supposed to be used like this - ItemsRepeater inside of ScrollViewer.

It's just a ScrollViewer on top of the ItemRepeater, I don't know why the code doesn't show the ScrollViewer tag, but the close-tag of it is there (last line of code).

> It always virtualizes, when it can.

When it cannot? maybe I'm doing something wrong.

urgh, I'm running out of options, do you have another idea?
Thanks in advance.

**Reply by maxkatz6** [reply]
@wieslawsoltes was able to create a small example of excel-like table with 10_000x1_000_000 items https://github.com/wieslawsoltes/VirtualDataGridDemo
Although it seems to crash anyway with horizontal scrolling sometimes

**Reply by jvulcan** [reply]
Thank you for the example project, at least for me it doesn't crash when scrolling horizontally (I updated the AvaloniaUI nuget packet) and it looks like a good template to start. I pressume that is the 'best' solution without entering in heavy optimization. The only issue is that the horizontal scrollbar arrow buttons don't work, but the vertical one works, pretty strange. There's no ISSUE to post a bug, that's why I'm giving the insight here.

I think this example is the best solution to start so I will check it as a solution for now. If I have other walls I will make a different topic.
Thank you kindly!

**Reply by wieslawsoltes** [reply]
I have fixed scroll crash in the meantime. Enabled issues in repo. The code needs proper refactoring into templated  control and lot's more work anyway.

**Comment by grokys** [comment]
How have you determined that it's not virtualizing? When you have a single column, taking a look in DevTools, how many visual children of the `ItemsRepeater` are created?

**Reply by jvulcan** [reply]
Hi,
In the first state (where only the first column is drawn and the rest is invisible) the Visual Count is 16
![VisualCount](https://user-images.githubusercontent.com/581282/135758234-f50842c6-2d69-4d4b-8164-ae1e34b1ece2.png)

After scrolling down and then appearing out of the blue the Visual Count jumps to 521
![VisualCount_after](https://user-images.githubusercontent.com/581282/135758261-a9014157-6ef1-49de-b4cb-47be302498e4.png)

The Count should be 15 (Visible Alts) * 6 (Visible Crits) = 90 elements + another row and column.
